File: api-node/draftKings_api/controllers/tacticController.ts
import { Request, Response } from "express";
import { z } from "zod";
import { AiTacticService } from "../services/aiTacticService";

const aiTacticService = new AiTacticService();

// Validamos el body que envía el frontend
const requestBodySchema = z.object({
  positions: z.record(z.string(), z.string().nullable()),
});

export const getAiRecommendations = async (req: Request, res: Response) => {
  try {
    // 1. Validar Body
    const parsedBody = requestBodySchema.safeParse(req.body);

    if (
      !parsedBody.success ||
      Object.keys(parsedBody.data.positions).length === 0
    ) {
      return res.status(400).json({
        message:
          "Bad Request. El formato del mapa de posiciones es inválido o no se han enviado datos.",
      });
    }

    // 2. Obtener recomendaciones
    const { positions } = parsedBody.data;
    const aiResponse = await aiTacticService.getRecommendations(positions);

    // 3. Respuesta OK (El parser de LangChain ya nos asegura que es el JSON correcto)
    return res.status(200).json(aiResponse);
  } catch (err: any) {
    if (err.message === "NO_EMPTY_POSITIONS") {
      return res.status(400).json({
        message: "Bad Request. No hay posiciones vacías para recomendar.",
      });
    }

    if (err.message === "AI_SERVICE_ERROR") {
      return res.status(503).json({
        message:
          "Service Unavailable. Error de comunicación o timeout con el proveedor de IA (Groq).",
      });
    }

    return res
      .status(500)
      .json({ message: "Unknown Error", error: err.message });
  }
};

File: api-spring/eureka.client.player/src/main/java/draftkings/eureka/client/player/controller/TacticController.java
package draftkings.eureka.client.player.controller;

import draftkings.eureka.client.player.dto.TacticRecommendationRequestDTO;
import draftkings.eureka.client.player.dto.TacticRecommendationResponseDTO;
import draftkings.eureka.client.player.service.AiTacticService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.Map;

@RestController
@RequestMapping("/api/tactics")
public class TacticController {

    private final AiTacticService aiTacticService;

    public TacticController(AiTacticService aiTacticService) {
        this.aiTacticService = aiTacticService;
    }

    @PostMapping("/recommendations")
    public ResponseEntity<?> getAiRecommendations(@RequestBody TacticRecommendationRequestDTO request) {
        // Validación: Bad Request si el cuerpo o el mapa están vacíos
        if (request == null || request.positions() == null || request.positions().isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("message",
                            "El formato del mapa de posiciones es inválido o no se han enviado datos."));
        }

        try {
            TacticRecommendationResponseDTO response = aiTacticService.getRecommendations(request.positions());
            return ResponseEntity.ok(response);

        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("message", "No hay posiciones vacías para cubrir en esta alineación."));

        } catch (IllegalStateException e) {
            // Captura el error de timeout/comunicación con Groq (503)
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(Map.of("message",
                            "Error de comunicación o timeout con el proveedor del servicio de Inteligencia Artificial."));

        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Unknown Error"));
        }
    }
}